File: frontend/src/lib/logger/logger.ts
import { appendFile, mkdir, stat } from "node:fs/promises";
import path from "node:path";

//define niveis de log
type LogLevel = "INFO" | "WARNING" | "ERROR" | "CRITICAL";

//define cores para cada nivel de log
const COLORS: Record<LogLevel, string> = {
    INFO: "\x1b[32m",
    WARNING: "\x1b[33m",
    ERROR: "\x1b[31m",
    CRITICAL: "\x1b[35m",
};

//cor da parte da mensagem da linha de log
const RESET = "\x1b[0m";


function getLogDir() { //retorna dir da pasta de logs
    return process.env.AGENT_LOG_DIR || "logs";
}

function getLogFilePath() { //retona caminho do arquivo de logs
    const today = new Date().toISOString().split("T")[0];

    return path.join(process.cwd(), getLogDir(), `agent-${today}.log`);
}

//formata datetime para cada linha de log
function formatDateTime() {
    const now = new Date();

    const day = String(now.getDate()).padStart(2, "0");
    const month = String(now.getMonth() + 1).padStart(2, "0");
    const year = now.getFullYear();

    const hours = String(now.getHours()).padStart(2, "0");
    const minutes = String(now.getMinutes()).padStart(2, "0");

    return `${day}/${month}/${year} ${hours}:${minutes}`;
}

//formatacao de cores para cada parte da linha de log
function colorizeLevel(level: LogLevel) {
    return `${COLORS[level]}[${level}]${RESET}`;
}

//retorna o arquivo que chamou o logger
function getCallerPath() {
    const stack = new Error().stack;    //retorna pilha de chamadas 
                                        // stack mostra quais funcoes foram chamadas (arquivo linha e coluna) - usa como trace das funcoes chamadas

    if (!stack) {
        return "unknown"; // caso n encontre o arquivo retorna unknow
    }

    const stackLines = stack.split("\n"); //quebra string em array de linhas
    const callerLine = stackLines[4] || stackLines[3] || ""; //tenta pegar a linha do arquivo que chamou o logger

    const match = // 2 opcoes de regex para pegar caminho da funcao que chama o logger
        callerLine.match(/\((.*):\d+:\d+\)/) ||
        callerLine.match(/at (.*):\d+:\d+/);

    if (!match) {
        return "unknown";
    }

    const absolutePath = match[1];

    return path //retorna caminho relativo
        .relative(process.cwd(), absolutePath)
        .replaceAll("\\", "/");
}


//garante que arquivo de logger tenha linha inicial
async function ensureLogStarted(filePath: string) {
    try { //verifica se o arquivo existe e se tem linha inicial
        const fileStat = await stat(filePath);

        if (fileStat.size > 0) {
            return;
        }
    } catch {
    // Se o arquivo ainda não existe, ele será criado abaixo.
    }
    //se nao tiver arquivo, cria abaixo e adiciona linha inicial no arquivo de log
    const startLine = `========== LOG INICIADO EM ${formatDateTime()} ==========\n`;

    await appendFile(filePath, startLine, "utf-8");
}

//funcao de escrita do logger
async function write(level: LogLevel, message: string) {
    // pega caminhos (caminho de arquivo e diretorio)
    const filePath = getLogFilePath();
    const dirPath = path.dirname(filePath);

    //tenta encontrar aqrquivo, se nao existir, cria diretorio, arquivo e adiciona linha inicial 
    await mkdir(dirPath, { recursive: true });
    await ensureLogStarted(filePath);

    //pega caminho da funcao que chamou o logger e pega datetime
    const callerPath = getCallerPath();
    const dateTime = formatDateTime();

    //define formato de linha do log
    const fileLine = `[${callerPath}] [${level}] ${dateTime} - ${message}\n`;

    const consoleLine = `[${callerPath}] ${colorizeLevel(
        level
    )} ${dateTime} - ${message}`;

    console.log(consoleLine); //imprime no terminal

    await appendFile(filePath, fileLine, "utf-8"); //adiciona linha de log no arquivo logger
}


//funcoes prontas para escrever log para cada nivel (padronizado)
export const logger = {
    info(message: string) {
        return write("INFO", message);
    },

    warning(message: string) {
        return write("WARNING", message);
    },

    error(message: string) {
        return write("ERROR", message);
    },

    critical(message: string) {
        return write("CRITICAL", message);
    },
};

File: backend/src/infrastructure/logger/logger.go
package logger

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

type LogLevel string

// define constantes para niveis de log
const (
	INFO     LogLevel = "INFO"
	WARNING  LogLevel = "WARNING"
	ERROR    LogLevel = "ERROR"
	CRITICAL LogLevel = "CRITICAL"
)

// define constantes para cores de niveis de log
const (
	colorReset   = "\033[0m"
	colorGreen   = "\033[32m"
	colorYellow  = "\033[33m"
	colorRed     = "\033[31m"
	colorPurple  = "\033[35m"
)

func getLogDir() string {
	logDir := os.Getenv("BACKEND_LOG_DIR")
	if logDir == "" {
		logDir = "logs"
	}

	return logDir
}

func getLogFilePath() string { // monta caminho de arquivo com nome backend-data.log
	today := time.Now().Format("2006-01-02")
	fileName := fmt.Sprintf("backend-%s.log", today)

	return filepath.Join(getLogDir(), fileName)
}

func getFormattedTime() string { //formata datetime para linhas de log
	return time.Now().Format("02/01/2006 15:04")
}


func colorizeLevel(level LogLevel) string {
	levelText := fmt.Sprintf("[%s]", level)

	switch level {
	case INFO:
		return colorGreen + levelText + colorReset
	case WARNING:
		return colorYellow + levelText + colorReset
	case ERROR:
		return colorRed + levelText + colorReset
	case CRITICAL:
		return colorPurple + levelText + colorReset
	default:
		return levelText
	}
}

func getCallerPath() string {
	_, file, _, ok := runtime.Caller(3)
	if !ok {
		return "unknown"
	}

	workingDir, err := os.Getwd()
	if err != nil {
		return file
	}

	relativePath, err := filepath.Rel(workingDir, file)
	if err != nil {
		return file
	}

	return strings.ReplaceAll(relativePath, "\\", "/")
}

func ensureLogFileStarted(filePath string) {
	info, err := os.Stat(filePath)

	if err == nil && info.Size() > 0 {
		return
	}

	file, err := os.OpenFile(filePath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
	if err != nil {
		log.Println("Erro ao criar arquivo de log:", err)
		return
	}
	defer file.Close()

	startLine := fmt.Sprintf(
		"========== LOG INICIADO EM %s ==========\n",
		time.Now().Format("02/01/2006 15:04:05"),
	)

	if _, err := file.WriteString(startLine); err != nil {
		log.Println("Erro ao escrever início do log:", err)
	}
}

func write(level LogLevel, message string) {
	filePath := getLogFilePath()
	dirPath := filepath.Dir(filePath)

	if err := os.MkdirAll(dirPath, os.ModePerm); err != nil {
		log.Println("Erro ao criar diretório de logs:", err)
		return
	}

	ensureLogFileStarted(filePath)

	callerPath := getCallerPath()
	dateTime := getFormattedTime()

	fileLine := fmt.Sprintf(
		"[%s] [%s] %s - %s\n",
		callerPath,
		level,
		dateTime,
		message,
	)

	//retorna string formatada da linha de log
	consoleLine := fmt.Sprintf(
		"[%s] %s %s - %s",
		callerPath,
		colorizeLevel(level),
		dateTime,
		message,
	)

	//retorna linha formatada no stdout
	fmt.Println(consoleLine)


	file, err := os.OpenFile(filePath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
	if err != nil {
		log.Println("Erro ao abrir arquivo de log:", err)
		return
	}
	defer file.Close()

	if _, err := file.WriteString(fileLine); err != nil {
		log.Println("Erro ao escrever arquivo de log:", err)
	}
}


//cria funcoes para escrever log em niveis de logs diferentes (padronizado)

func Info(message string) {
	write(INFO, message)
}

func Warning(message string) {
	write(WARNING, message)
}

func Error(message string) {
	write(ERROR, message)
}

func Critical(message string) {
	write(CRITICAL, message)
}